File: services/anchor-service/internal/merkle.go
package internal

import (
    "crypto/sha256"
    "encoding/hex"
)

// combineRoots does a pairwise combine of hex-encoded roots and returns hex root
func CombineRootsHex(roots []string) (string, error) {
    if len(roots) == 0 {
        return "", nil
    }
    layer := make([][]byte, 0, len(roots))
    for _, r := range roots {
        // trim 0x
        if len(r) >= 2 && r[:2] == "0x" {
            r = r[2:]
        }
        b, err := hex.DecodeString(r)
        if err != nil {
            return "", err
        }
        layer = append(layer, b)
    }

    for len(layer) > 1 {
        next := make([][]byte, 0, (len(layer)+1)/2)
        for i := 0; i < len(layer); i += 2 {
            left := layer[i]
            var right []byte
            if i+1 < len(layer) {
                right = layer[i+1]
            } else {
                right = left
            }
            concat := append(left, right...)
            h := sha256.Sum256(concat)
            next = append(next, h[:])
        }
        layer = next
    }
    return hex.EncodeToString(layer[0]), nil
}

File: services/auditor-verifier/src/merkle.ts
import crypto from 'crypto';

/**
 * Compute leaf hash from hex string (supports "0x" prefix)
 */
export function hashLeaf(hexRoot: string): Buffer {
  const h = hexRoot.startsWith('0x') ? hexRoot.slice(2) : hexRoot;
  // const bytes = Buffer.from(h, 'hex');
  // return crypto.createHash('sha256').update(bytes).digest();
  return Buffer.from(h, 'hex');
}

/**
 * Combine two buffers: hash( left || right )
 */
function hashPair(a: Buffer, b: Buffer): Buffer {
  // deterministic order: lexicographic - many merkle schemes do this, adjust if your scheme differs
  const left = Buffer.compare(a, b) <= 0 ? a : b;
  const right = Buffer.compare(a, b) <= 0 ? b : a;
  return crypto.createHash('sha256').update(Buffer.concat([left, right])).digest();
}

/**
 * Combine array of hex roots -> final hex root (no 0x prefix)
 */
export function combineRootsHex(hexRoots: string[]): string {
  if (!Array.isArray(hexRoots) || hexRoots.length === 0) {
    throw new Error('no roots');
  }
  let layer = hexRoots.map(r => hashLeaf(r));
  while (layer.length > 1) {
    const next: Buffer[] = [];
    for (let i = 0; i < layer.length; i += 2) {
      if (i + 1 === layer.length) {
        // odd: carry up (hash of single child)
        next.push(layer[i]);
      } else {
        next.push(hashPair(layer[i], layer[i + 1]));
      }
    }
    layer = next;
  }
  return layer[0].toString('hex'); // no 0x
}
